src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1735,6 +1717,62 @@ void MarkField (FieldDefinition field, in DependencyInfo reason, in MessageOrigi
 			}
 		}
 
+		bool ShouldWarnForReflectionAccessToCompilerGeneratedCode (FieldDefinition field)
+		{
+			if (!CompilerGeneratedState.IsNestedFunctionOrStateMachineMember (field))
+				return false;
+
+			// No need to warn if it's already covered by the Requires attribute or explicit annotations on the field.
+			if (Annotations.DoesFieldRequireUnreferencedCode (field, out _) || Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (field))
+				return false;
+
+			// Only warn for types which are interesting for dataflow. Note that this does
+			// not include integer types, even though we track integers in the dataflow analysis.
+			// Technically we should also warn for integer types, but this leads to more warnings
+			// for example about the compiler-generated "state" field for state machine methods.
+			// This should be ok because in most cases the state machine types will also have other
+			// hoisted locals that produce warnings anyway when accessed via reflection.
+			return Annotations.FlowAnnotations.IsTypeInterestingForDataflow (field.FieldType);
+		}
+
+		void ProcessAnalysisAnnotationsForField (FieldDefinition field, DependencyKind dependencyKind, in MessageOrigin origin)
+		{
+			if (origin.Provider != ScopeStack.CurrentScope.Origin.Provider) {
+				Debug.Assert (dependencyKind == DependencyKind.DynamicallyAccessedMemberOnType ||
+					(origin.Provider is MethodDefinition originMethod && CompilerGeneratedState.IsNestedFunctionOrStateMachineMember (originMethod)));
+			}
+
+			if (dependencyKind == DependencyKind.DynamicallyAccessedMemberOnType) {
+				ReportWarningsForTypeHierarchyReflectionAccess (field, origin);
```

Comment:
Does this mean that we will get more noise when using type annotations?
This code path doesn't use the heuristic to determine which fields to warn about - and it can still effectively apply `All` to a type (for example).

Reply:
Perfectly OK to try to create an issue/postpone - since I think we have the problem already.

Reply:
Well - please add a test for this at least.